library/spdm_requester_lib/libspdm_req_subscribe_event_types.c

```diff
@@ -0,0 +1,182 @@
+    if (LIBSPDM_STATUS_IS_ERROR(status)) {
+        libspdm_release_sender_buffer (spdm_context);
+        return status;
+    }
+
+    libspdm_release_sender_buffer(spdm_context);
+    spdm_request = (void *)spdm_context->last_spdm_request;
+
+    /* -=[Receive Response Phase]=- */
+    status = libspdm_acquire_receiver_buffer(spdm_context, &message_size, (void **)&message);
+    if (LIBSPDM_STATUS_IS_ERROR(status)) {
+        return status;
+    }
+    LIBSPDM_ASSERT (message_size >= transport_header_size);
+    spdm_response = (void *)(message);
+    spdm_response_size = message_size;
+
+    status = libspdm_receive_spdm_response(
+        spdm_context, &session_id, &spdm_response_size, (void **)&spdm_response);
+    if (LIBSPDM_STATUS_IS_ERROR(status)) {
+        goto receive_done;
+    }
+
+    /* -=[Validate Response Phase]=- */
+    if (spdm_response_size != sizeof(spdm_subscribe_event_types_ack_response_t)) {
+        status = LIBSPDM_STATUS_INVALID_MSG_SIZE;
+        goto receive_done;
+    }
+    if (spdm_response->header.spdm_version != spdm_request->header.spdm_version) {
+        status = LIBSPDM_STATUS_INVALID_MSG_FIELD;
+        goto receive_done;
+    }
+    if (spdm_response->header.request_response_code == SPDM_ERROR) {
+        status = libspdm_handle_error_response_main(
+            spdm_context, &session_id,
+            &spdm_response_size, (void **)&spdm_response,
+            SPDM_SUBSCRIBE_EVENT_TYPES, SPDM_SUBSCRIBE_EVENT_TYPES_ACK);
+        if (LIBSPDM_STATUS_IS_ERROR(status)) {
+            goto receive_done;
+        }
+    } else if (spdm_response->header.request_response_code != SPDM_SUBSCRIBE_EVENT_TYPES_ACK) {
+        status = LIBSPDM_STATUS_INVALID_MSG_FIELD;
+        goto receive_done;
+    }
+
+receive_done:
+    libspdm_release_receiver_buffer(spdm_context);
+
+    return status;
+}
+
+libspdm_return_t libspdm_subscribe_event_types(void *spdm_context,
+                                               uint32_t session_id,
+                                               uint8_t subscribe_event_group_count,
+                                               uint32_t subscribe_list_len,
+                                               void *subscribe_list)
+{
+    size_t retry;
+    uint64_t retry_delay_time;
+    libspdm_return_t status;
+    libspdm_context_t *context;
+
+    context = spdm_context;
+    context->crypto_request = true;
+    retry = context->retry_times;
+    retry_delay_time = context->retry_delay_time;
+    do {
+        status = try_libspdm_subscribe_event_types(context,
+                                                   session_id,
+                                                   subscribe_event_group_count,
+                                                   subscribe_list_len,
+                                                   subscribe_list);
+        if ((status != LIBSPDM_STATUS_BUSY_PEER) || (retry == 0)) {
```

Comment:
Don't need the `(retry == 0)` anymore. See https://github.com/DMTF/libspdm/pull/2815.